### PR TITLE
Cleanup the lexer

### PR DESCRIPTION
**Pull request by teo-tsirpanis**
This PR "cleans-up" the files `fslexast.fs` - and to a lesser extent, `Lexing.fs` -.

Some of the changes might improve performance, like [those on the first commit](d85d0e813c66fbb6ba9e34c407e33f3364d5c429).

Other changes include, but are not limited to removing unneeded trailing semicolons, and taking advantage of list comprehensions.

### CODE REVIEW

File: src/FsLexYacc.Runtime/Lexing.fs [hunk 1]
```diff
@@ -264,10 +264,10 @@ namespace FSharp.Text.Lexing
             LexBuffer<char>.FromChars (s.ToCharArray())
 
         static member FromTextReader (tr:System.IO.TextReader) : LexBuffer<char> = 
-           LexBuffer<char>.FromFunction(tr.Read) 
+           LexBuffer<char>.FromReadFunctions(Some tr.Read, Some (tr.ReadAsync >> Async.AwaitTask))
 
         static member FromBinaryReader (br:System.IO.BinaryReader) : LexBuffer<byte> = 
-           LexBuffer<byte>.FromFunction(br.Read) 
+           LexBuffer<byte>.FromFunction(br.Read)
```
Comment (teo-tsirpanis) [inline src/FsLexYacc.Runtime/Lexing.fs]: Why does that function even exist? The library was supposed to parse _text files_; not _binary_ ones.

At least, we could mark it as `Obsolete`.

File: src/FsYacc/AssemblyInfo.fs [hunk 2]
```diff
@@ -5,13 +5,13 @@ open System.Reflection
 [<assembly: AssemblyTitleAttribute("FsYacc")>]
 [<assembly: AssemblyProductAttribute("FsLexYacc")>]
 [<assembly: AssemblyDescriptionAttribute("FsLex/FsYacc lexer/parser generation tools")>]
-[<assembly: AssemblyVersionAttribute("9.0.3")>]
-[<assembly: AssemblyFileVersionAttribute("9.0.3")>]
+[<assembly: AssemblyVersionAttribute("9.1.0")>]
```
Comment (teo-tsirpanis) [inline src/FsYacc/AssemblyInfo.fs]: These files could be ignored by Git. The commits would have been much less cluttered.

File: src/FsLex/fslexast.fs [hunk 3]
```diff
@@ -18,51 +17,50 @@ type Alphabet = uint32
...
- dict 
-  [| "Pe", System.Globalization.UnicodeCategory.ClosePunctuation; // (Pe)
-    "Pc", System.Globalization.UnicodeCategory.ConnectorPunctuation; // (Pc)
-    "Cc", System.Globalization.UnicodeCategory.Control; // (Cc)
-    "Sc", System.Globalization.UnicodeCategory.CurrencySymbol; // (Sc)
-    "Pd", System.Globalization.UnicodeCategory.DashPunctuation; // (Pd)
-    "Nd", System.Globalization.UnicodeCategory.DecimalDigitNumber; // (Nd)
-    "Me", System.Globalization.UnicodeCategory.EnclosingMark; // (Me)
-    "Pf", System.Globalization.UnicodeCategory.FinalQuotePunctuation; // (Pf)
-    "Cf", enum 15; //System.Globalization.UnicodeCategory.Format; // (Cf)
```
Comment (teo-tsirpanis) [inline src/FsLex/fslexast.fs]: Was there a specific reason for that? The use of `enum 15` did stand out.

File: src/FsLex/fslexast.fs [hunk 4]
```diff
@@ -74,87 +72,87 @@ let numLowUnicodeChars = 128
...
+      Transitions: Dictionary<Alphabet, NfaNode list>
       Accepted: (int * int) list }
 
-type DfaNode = 
-    { Id: int;
-      Name: string;
-      mutable Transitions: (Alphabet * DfaNode) list;
+type DfaNode =
+    { Id: int
+      Name: string
```
Comment (teo-tsirpanis) [inline src/FsLex/fslexast.fs]: Why do DFA nodes have _names_? Aren't they supposed to be called "State 1", "State 2" etc.?

File: src/FsLex/fslexast.fs [hunk 5]
```diff
@@ -165,244 +163,236 @@ let AddToMultiMap (trDict:MultiMap<_,_>) a b =
...
     member x.IsEmpty = (s.Length = 0)
-    member x.Iterate f = s |> Array.iter f
+    member x.Iterate f = Array.iter f s
 
 type NodeSetSet = Set<NfaNodeIdSet>
 
-let newDfaNodeId = 
-    let i = ref 0 
+let newDfaNodeId =
+    let i = ref 0
```
Comment (teo-tsirpanis) [inline src/FsLex/fslexast.fs]: All these `ref` uses could be replaced with `mutable`.

**Comment by dsyme** [comment]
Could you fix the merge conflicts please, then @dsyme me again?  thanks

**Comment by teo-tsirpanis** [comment]
OK, @dsyme.

File: .gitignore [hunk 6]
```diff
@@ -198,3 +198,6 @@ tests/fsyacc/test1compat.exe.config
 tests/fsyacc/test2.exe.config
 tests/fsyacc/test2compat.exe.config
 .paket
+
+.idea/
+.vs/
```
Comment (sergey-tihon) [inline .gitignore]: @teo-tsirpanis  can you please revert this in PR? this change already merged into master

**Pull request merged**